### PR TITLE
feat(excelsheet): add support for multi-dtype columns

### LINKED ISSUE #160: dtype resolution is wrong for multi-dtype columns
Type coercion is currently done on the first cell of the column only. It should be done on the entire column. This needs to be addressed without impacting performance too much.

_Originally posted by @lukapeschke in https://github.com/ToucanToco/fastexcel/issues/158#issuecomment-1914324743_

### PR DESCRIPTION
**Pull request by lukapeschke**
closes #160

### CODE REVIEW

**Comment by lukapeschke** [comment]
# Performance impact

As expected, this has a small performance impact time-wise. However, the cost seems acceptable to me. If it seems unacceptable to others, I'm open for a refactoring to make this optional.

The good news is that there is no impact memory-wise. However, iterating over all data in the sheet to build the schema appears to cost some time (note the flat memory usage that appears in the charts just after the first usage spike).

Some ideas that could help to mitigate this:
* Allowing the dtypes to be specified (as suggested in #158), which would allow us to skip the dtype guessing step entirely
* Allowing to specify a maximal number of rows to sample for dtype guessing (for example, on my 280k rows spreadsheet, reading the first 1000 would be sufficient)

# Benchmarks
Using the following script, here are the results of benchmarks on my machine (sheet of 41 columns x ~280k rows):

```python
import argparse

import fastexcel


def get_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("file")
    return parser.parse_args()


def main():
    args = get_args()
    excel_file = fastexcel.read_excel(args.file)
    for sheet_name in excel_file.sheet_names:
        excel_file.load_sheet_by_name(sheet_name).to_arrow()


if __name__ == "__main__":
    main()
```

## Before
before

## After
after

## After, without the tweak in `create_string_array`
without_match

**Comment by lukapeschke** [comment]
cc @ldacey @deanm0000 @alexander-beedie in case you want to have a look :)

**Comment by ldacey** [comment]
Nice - yeah, being able to cast the type ahead of time or change the number of rows used for inference could be useful. I have raised an issue for polars before because a column which looked like an integer actually had some text values beyond the default row inference. In this case it was the 622nd row, but you can imagine a large file that has been sorted where all values with letters or pure numbers are grouped at the bottom of the file.

```python
import polars as pl
import tempfile

case_numbers = [f"{i:06d}" for i in range(1, 1001)]
test = ["test" for _ in range(1, 1001)]

df = pl.DataFrame({'test': test, 'case_number': case_numbers})

df[622, "case_number"] = "CASE-NO-0A60"

with tempfile.NamedTemporaryFile(delete=True, suffix=".csv") as temp_file:
    filename = temp_file.name
    df.write_csv(filename)
    pl.read_csv(filename, n_rows=1, infer_schema_length=623)
```

**Comment by PrettyWood** [comment]
I'll review when I'm back from holidays @lukapeschke 
Super glad fastexcel had been added to polars ecosystem as we first planned ;)
Great job

**Comment by lukapeschke** [comment]
@PrettyWood sure no worries :slightly_smiling_face: I've released 0.8.0 in the meantime so that people can try out the windows wheels

File: src/utils/arrow.rs [hunk 1]
```diff
@@ -53,13 +102,64 @@ pub(crate) fn arrow_schema_from_column_names_and_range(
...
+    // pure bool
+    #[case(0, 2, ArrowDataType::Boolean)]
+    // pure int
+    #[case(2, 3, ArrowDataType::Int64)]
+    // pure float
+    #[case(3, 4, ArrowDataType::Float64)]
+    // pure string
+    #[case(4, 5, ArrowDataType::Utf8)]
+    // pure int + float
+    #[case(2, 4, ArrowDataType::Float64)]
```
Comment (PrettyWood) [inline src/utils/arrow.rs:147]: Note for myself: add int + null

Comment (lukapeschke) [reply src/utils/arrow.rs:149]: [025bdc1](https://github.com/ToucanToco/fastexcel/pull/164/commits/025bdc1af50fbc153655ba6a67808c19ae66e5dd)

**Comment by lukapeschke** [comment]
@PrettyWood last changes add:
* Merge with main (calamine 0.24)
* Added support for bools for multi-dtype columns
* Adapted to recent calamine changes (stuff related to `ExcelDateTime`)
* Added `schema_sample_rows` param

File: src/utils/arrow.rs [hunk 2]
```diff
@@ -34,6 +36,70 @@ fn get_arrow_column_type(data: &Range<CalData>, row: usize, col: usize) -> Resul
...
+}
+
+fn int_types() -> &'static HashSet<ArrowDataType> {
+    INT_TYPES_CELL.get_or_init(|| HashSet::from([ArrowDataType::Int64, ArrowDataType::Boolean]))
+}
+
+fn string_types() -> &'static HashSet<ArrowDataType> {
+    STRING_TYPES_CELL.get_or_init(|| {
+        HashSet::from([
+            ArrowDataType::Int64,
```
Comment (PrettyWood) [inline src/utils/arrow.rs:60]: the case boolean + string is impossible?

**Review by PrettyWood — COMMENTED** [review]
LGTM apart one question

File: src/utils/arrow.rs (hunk 2 above)
Comment (lukapeschke) [reply src/utils/arrow.rs:60]: It is probably possible, but string to bool conversion is not supported by calamine

**Review by PrettyWood — APPROVED** [review]

**Pull request merged**